x/dex/types/tick_liquidity.go

```diff
@@ -1,17 +1,73 @@
 package types
 
-import (
-	sdk "github.com/cosmos/cosmos-sdk/types"
-)
+func (t TickLiquidity) TokenIn() string {
+	switch liquidity := t.Liquidity.(type) {
+	case *TickLiquidity_LimitOrderTranche:
+		return liquidity.LimitOrderTranche.TokenIn
 
-func (t TickLiquidity) HasToken() bool {
-	return t.HasLPReserves() || t.HasActiveLimitOrders()
+	case *TickLiquidity_PoolReserves:
+		return liquidity.PoolReserves.TokenIn
+	default:
+		panic("Tick does not contain valid liqudityType")
+	}
 }
 
-func (t TickLiquidity) HasLPReserves() bool {
-	return t.LPReserve != nil && t.LPReserve.GT(sdk.ZeroInt())
+func (t TickLiquidity) PairId() *PairId {
+	switch liquidity := t.Liquidity.(type) {
+	case *TickLiquidity_LimitOrderTranche:
+		return liquidity.LimitOrderTranche.PairId
+
+	case *TickLiquidity_PoolReserves:
+		return liquidity.PoolReserves.PairId
+	default:
+		panic("Tick does not contain valid liqudityType")
+	}
 }
 
-func (t TickLiquidity) HasActiveLimitOrders() bool {
-	return t.LimitOrderTranche != nil && t.LimitOrderTranche.ReservesTokenIn.GT(sdk.ZeroInt())
+func (t TickLiquidity) TickIndex() int64 {
+	switch liquidity := t.Liquidity.(type) {
+	case *TickLiquidity_LimitOrderTranche:
+		return liquidity.LimitOrderTranche.TickIndex
+
+	case *TickLiquidity_PoolReserves:
+		return liquidity.PoolReserves.TickIndex
+	default:
+		panic("Tick does not contain valid liqudityType")
+	}
+}
+
+func (t TickLiquidity) LiquidityType() string {
+	switch t.Liquidity.(type) {
+	case *TickLiquidity_LimitOrderTranche:
+		return LiquidityTypeLimitOrder
+
+	case *TickLiquidity_PoolReserves:
+		return LiquidityTypePoolReserves
+	default:
+		panic("Tick does not contain valid liqudityType")
+	}
+}
+
+func (t TickLiquidity) LiquidityIndex() uint64 {
+	switch liquidity := t.Liquidity.(type) {
+	case *TickLiquidity_LimitOrderTranche:
+		return liquidity.LimitOrderTranche.TrancheIndex
+
+	case *TickLiquidity_PoolReserves:
+		return liquidity.PoolReserves.Fee
+	default:
+		panic("Tick does not contain valid liqudityType")
+	}
+}
+
+func (t TickLiquidity) HasToken() bool {
```

Comment:
My preferred approach here is to define a separate ITickLiquidity interface which both LimitOrderTranche and PoolReserves implement separately and then whenever that interface is needed, define the object's type as that. Then the compiler does the dirty work of directing the function call.

Reply:
Opting to go for explicit references to each type rather than polymorphism. This makes more sense since 95% of the time we know which type we are dealing with so it doesn't make sense to use interfaces. Also polymorphic approach results in a lot of ugly type coercion.